### PR TITLE
Disable chroma subsampling in JPEGCodec

### PR DESCRIPTION
**Pull request by iwbh15**
Follow up to PR https://github.com/ome/ome-codecs/pull/26
 
see https://github.com/ome/ome-codecs/pull/26#issuecomment-1456675017 :

There is another option that has a decisive influence on the Jpeg image quality: '**Chroma Subsampling**'.
Especially with biological/microscopic images in RGB format (transmitted light or fluorescence) chroma subsampling leads to intolerable image errors.
Chroma subsampling should therefore - in my view - be deactivatable in libraries like BioFormats.

Unfortunately, access to the 'Chroma Sampling' property is not configurable via the 'ImageWriteParam' (like e.g. the JpegQuality).
Instead, it can be accessed via the 'IIOMetadata'.

Unfortunately, so far have not been able to find a direct and simple description.
But it seems to exist in some form in the "JPEG Metadata Format Specification and Usage Notes".

In ImageJ, access to the 'Chroma Sampling' property is already integrated.
see https://imagej.nih.gov/ij/notes.html
_1.52s 10 December 2019
    Thanks to Peter Haub, chroma subsampling is disabled when saving in JPEG format if the Quality setting (Edit>Options>Input/Output) is 90 or higher, resulting in higher quality images but larger file sizes._

For example, chroma subsampling can be disabled with the following macro command:
JpegWriter.disableChromaSubsampling(true);
By default, 'Chroma Sampling' is disabled in ImageJ when JpegQuality >= 90.

My source at that time was Apache/Shindig.
See:
https://github.com/imagej/ImageJ/blob/d82104f1c55290af56d5afc29004cf8298055067/ij/plugin/JpegWriter.java#L86
_// Disable JPEG chroma subsampling
// http://svn.apache.org/repos/asf/shindig/trunk/java/gadgets/src/main/java/org/apache/shindig/gadgets/rewrite/image/BaseOptimizer.java
// http://svn.apache.org/repos/asf/shindig/trunk/java/gadgets/src/main/java/org/apache/shindig/gadgets/rewrite/image/JpegImageUtils.java_

There is a test script from Wayne here https://imagej.nih.gov/ij/macros/js/JpegQualityTests.js .
It can be called in the ImageJ menu under:  Help>Examples>Java Scripts> JPEG Quality Plot

I think it would be worth and important to have the 'Chroma Sampling' property in mind, especially with the current code changes to support variable Jpeg quality.

Unfortunately I lack the overview of ome-codecs to elaborate this hint more concretely.

Here are some further and interesting links:
_https://stackoverflow.com/questions/14149739/disable-java-imageio-chroma-subsampling
https://stackoverflow.com/questions/72581932/why-is-the-color-of-my-image-changed-after-writing-it-as-a-jpg-file
https://github.com/libvips/libvips/issues/538
http://poynton.ca/PDFs/Chroma_subsampling_notation.pdf_

### CODE REVIEW

**Comment by iwbh15** [comment]
**Important Note:**
The change made in `ome.codecs.CodecOptions.java` must also be inserted into `loci.formats.codec.CodecOptions`.
Can somebody please create this change in bioformats.

This is because `loci.formats.codec.CodecOptions` will be wrapped to `ome.codecs.CodecOptions.java`
in WrappedCodec.java (see https://github.com/ome/bioformats/blob/971fb341745e5c7e4b0cbf8c2af59deacc73db75/components/formats-bsd/src/loci/formats/codec/WrappedCodec.java#L139)
(called from TiffCompression.java, see https://github.com/ome/bioformats/blob/971fb341745e5c7e4b0cbf8c2af59deacc73db75/components/formats-bsd/src/loci/formats/tiff/TiffCompression.java#L360)

**Comment by dgault** [comment]
Thanks @iwbh15 for getting this PR opened, we are currently preparing a Bio-Formats 6.14.0 release for next week. Once that is out we we will aim to get this reviewed and tested for a Bio-Formats 7.0.0 which will follow shortly after.

File: src/main/java/ome/codecs/JPEGCodec.java
Comment (melissalinkert) [inline src/main/java/ome/codecs/JPEGCodec.java:190]: I have not so far been able to create an example where the `SOF` node has 3 child nodes. I believe this is because `img` cannot have the relevant ColorSpace/ColorModel (see https://github.com/ome/ome-codecs/blob/master/src/main/java/ome/codecs/gui/AWTImageTools.java#L1077 and https://github.com/ome/ome-codecs/blob/master/src/main/java/ome/codecs/gui/AWTImageTools.java#L1101), but perhaps I misunderstand.

Note I tested with https://github.com/ome/bioformats/pull/4057 and the snippet therein. Is there an example that shows this code working as intended?

File: src/main/java/ome/codecs/JPEGCodec.java
Comment (iwbh15) [reply src/main/java/ome/codecs/JPEGCodec.java]: Inline usernames removed.

File: src/main/java/ome/codecs/JPEGCodec.java
Comment (iwbh15) [reply src/main/java/ome/codecs/JPEGCodec.java:123]: The presented version is the current implementation in ImageJ (slightly modified).
The idea was to automatically disable chroma subsampling above a certain JPEG quality. This idea is understandable and makes sense. Especially when it comes to the highest image quality, chroma subsampling is the most annoying. In addition it should be possible to enable an explicit on/off with higher priority.
How this option is used here is of course an OME design decision.
All this has nothing to do with the limitation of the quality range [0.25, 1.0].

File: src/main/java/ome/codecs/JPEGCodec.java
Comment (iwbh15) [reply src/main/java/ome/codecs/JPEGCodec.java:187]: **I)**
Basically, your test snippet works.

The problem is that the setting of the option `disableChromaSubsampling` is lost on the way from the snippet to the `JPEGCodec.java`.

Namely in the constructor
[public CodecOptions(CodecOptions options)](https://github.com/ome/bioformats/blob/3d4f0ca079a97070cbb647333e536f0f164d91a1/components/formats-api/src/loci/formats/codec/CodecOptions.java#L126)
and in the function
[private static void copyOptions(CodecOptions src, ome.codecs.CodecOptions dest)](https://github.com/ome/bioformats/blob/3d4f0ca079a97070cbb647333e536f0f164d91a1/components/formats-bsd/src/loci/formats/codec/WrappedCodec.java#L139C2-L139C82)

In both places the code must be supplemented as follows:
`dst.disableChromaSubsampling = src.disableChromaSubsampling;`
(of course also with corresponding declaration of the variable disableChromaSubsampling).

Alternatively, you can also temporarely hardcode the option disableChromaSubsampling in the JPEGCodec.java
`disableChromaSubsampling = true / false;`

**II)**
In the structured fake data (where R = G = B) the effect of the disabled chroma subsampling is not huge.
Therefore I have use the image `lena-std.tif` for my test (available in the [ImageJ example images zip archive](https://imagej.nih.gov/ij/download/sample-images.zip)).
(Of course you can use any other RGB tif image.)

I used a [modified snipped](https://github.com/ome/bioformats/pull/4057#issuecomment-1645068731) for the test.

The test shows that the resulting images (with / without chroma subsampling) are different and that the image saved **with** `disableChromaSubsampling` is closer to the original, as expected.

_(The resulting images can be opened correctly in ImageJ and QuPath. But it seems that Windows Preview and other viewers misinterpret the images as YCbCr. I could not find the necessary settings to turn this off. However, this is not relevant for the test.)_

**III)**
Apart from the basic functionality of the option `disableChromaSubsampling`, I was surprised by the fact that Tif images saved with JPEG compression (at high quality settings) are larger than the original uncompressed image.
Apparently, the image is saved line by line, which does not allow the full potential of JPEG compression to be realized. I have not investigated this circumstance further. But it should be considered to **localize** the cause of **this behavior**, because in this form JPEG compression is not really helpful.

_At this point, perhaps a word about the application:
WSI images - which are to a large extent RGB transmitted light images - are still saved by various manufacturers as Tif with JPEG compression. From my point of view it would be helpful if this format could also be written - regardless of the disadvantageous JPEG artifacts. It is precisely to minimize these artifacts that switching off chroma subsampling serves and to be able to create Tif images with JPEG compression with high quality._

**IV)**
As can be seen from the tests, the color representation does not have to be explicitly YCbCr.
The listed node conditions are also fulfilled with RGB representation.

File: src/main/java/ome/codecs/JPEGCodec.java
Comment (melissalinkert) [reply src/main/java/ome/codecs/JPEGCodec.java:123]: Line 122 sets `disableChromaSubsampling` to true if `jpegquality` is greater than 90. But the value of `jpegquality` in that line can never be greater than 1.0, correct? If so, that means that `disableChromaSubsampling` will always be set to false in line 122. If the intention is to automatically disable chroma subsampling above 90% quality, then the check needs to be changed to `jpegquality >= 0.9`.

Comment (iwbh15) [reply src/main/java/ome/codecs/JPEGCodec.java:123]: Query changed to `jpegquality >= 0.9`

File: src/main/java/ome/codecs/JPEGCodec.java
Comment (iwbh15) [reply src/main/java/ome/codecs/JPEGCodec.java:187]: Note:
" ... the image is saved line by line, which does not allow the full potential of JPEG compression to be realized. ..."

In order to investigate this issue further, I have temporarily modified the function `writeImage` in `loci.formats.tiff.TiffSaver.java.`

I have temporarily inserted 
```
	if (compression.getCodecName().equals("JPEG"))      
	   ifd.putIFDValue(IFD.ROWS_PER_STRIP, new long[] {8});
```
directly after the [line](https://github.com/ome/bioformats/blob/3d4f0ca079a97070cbb647333e536f0f164d91a1/components/formats-bsd/src/loci/formats/tiff/TiffSaver.java#L313)
`	compression = ifd.getCompression();
`

This modification significantly reduces the size of the Jpeg compressed file.
In my test (with jpegquality = 0.75 and disableChromaSubsampling = false) the file size reduced from **585 kB** to **83 kB**.

**Review by melissalinkert — APPROVED** [review]
Based on testing in https://github.com/ome/bioformats/pull/4057#issuecomment-1828296074, I think I'm OK with this now, particularly as the default behavior is unchanged. It would be good to have a second reviewer though before merging.

**Pull request merged**
